### PR TITLE
Neptune integration

### PR DESCRIPTION
**Pull request by jakubczakon**
# Before submitting

- [X] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [X] Did you read the [contributor guideline](https://github.com/williamFalcon/pytorch-lightning/blob/master/.github/CONTRIBUTING.md)?
- [] Did you make sure to update the docs?   
- [X] Did you write any new necessary tests?  

## What does this PR do?
Creates integration described in #642

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
I sure did :)

## Note
I had trouble finding how to set up local docs so I am not sure if those work property.
Also, I added note about logging to readme which links to docs which I am quite sure are not created.
So I presume there will be trouble here.

### CODE REVIEW

**Comment by jakubczakon** [comment]
It seems that the failed tests are related to `test_restore_module.py`
Specifically, the accuracy of loaded models is `< 0.5`
```
 AssertionError: this model is expected to get > 0.5 in test set (it got 0.4375)
```
Not sure how to proceed.
Is there any way you could help me with this?

File: pytorch_lightning/logging/neptune.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+        self.logger.experiment.whatever_neptune_supports(...)
+
+"""
+
+from logging import getLogger
+
+try:
+    import neptune
+except ImportError:
+    raise ImportError('Missing neptune package. Run pip install neptune-client')
```
Comment (Borda) [inline pytorch_lightning/logging/neptune.py]: pls add ` to the command

File: tests/test_logging.py [hunk 2]
```diff
@@ -193,6 +193,58 @@ def test_comet_pickle(tmpdir, monkeypatch):
     trainer2.logger.log_metrics({"acc": 1.0})
 
 
+def test_neptune_logger(tmpdir):
+    """Verify that basic functionality of neptune logger works."""
+    tutils.reset_seed()
+
+    try:
```
Comment (Borda) [inline tests/test_logging.py]: remove this try/except

File: tests/test_logging.py [hunk 3]
```diff
@@ -193,6 +193,58 @@ def test_comet_pickle(tmpdir, monkeypatch):
...
+
+    print('result finished')
+    assert result == 1, "Training failed"
+
+
+def test_neptune_pickle(tmpdir):
+    """Verify that pickling trainer with neptune logger works."""
+    tutils.reset_seed()
+
+    try:
```
Comment (Borda) [inline tests/test_logging.py]: remove this try/except

File: pytorch_lightning/logging/neptune.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        """Set key-value pair as Neptune experiment property.
+
+        :param str key: Property key.
+        :param obj value: New value of a property.
+
+        """
+        self.experiment.set_property(key, value)
+
+    @rank_zero_only
+    def append_tags(self, tag, *tags):
```
Comment (Borda) [inline pytorch_lightning/logging/neptune.py]: this tag/tags as two variables sound quite strange, rather make it generic and just single var

File: pytorch_lightning/logging/neptune.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+        else:
+            self.experiment.log_metric(log_name, x=step, y=text)
+
+    @rank_zero_only
+    def log_image(self, log_name, image, step=None):
+        """Log image data in Neptune experiment
+
+        :param str log_name: The name of log, i.e. bboxes, visualisations, sample_images.
+        :param str|PIL.Image|matplotlib.figure.Figure image: The value of the log (data-point).
+        Can be one of the following types: PIL image, matplotlib.figure.Figure, path to image file (str)
```
Comment (Borda) [inline pytorch_lightning/logging/neptune.py]: add spaces at the beginning, so it is assigned to the paramaters

File: pytorch_lightning/logging/neptune.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+
+class NeptuneLogger(LightningLoggerBase):
+    def __init__(self, api_key=None, project_name=None, offline_mode=False,
+                 experiment_name=None, upload_source_files=None,
+                 params=None, properties=None, tags=None, **kwargs):
+        """Initialize a neptune.ml logger.
+        Requires either an API Key (online mode) or a local directory path (offline mode)
+
+        :param str|None api_key: Required in online mode. Neputne API token, found on https://neptune.ml.
+        Read how to get your API key here https://docs.neptune.ml/python-api/tutorials/get-started.html#copy-api-token.
```
Comment (Borda) [inline pytorch_lightning/logging/neptune.py]: add spaces at the beginning, so it is assigned to the parameters

File: pytorch_lightning/logging/neptune.py (hunk 4 above)
Comment (jakubczakon) [reply pytorch_lightning/logging/neptune.py]: The idea here is to support all the following options:

```python
logger.append_tags('resnet')
logger.append_tags('resnet', 'augmentations')
logger.append_tags(['resnet','augmentations'])
```

But I think for the simplicity I will go with the last option (list).

**Comment by Borda** [comment]
To the failing test, pls try increase number of training epochs, like a double...

**Comment by jakubczakon** [comment]
Ok on it (failing test) @Borda

**Comment by jakubczakon** [comment]
3 travis jobs worked but 2 still fail.
Could you please take a look @Borda 
https://travis-ci.org/williamFalcon/pytorch-lightning/jobs/629639792?utm_medium=notification&utm_source=github_status#L698

File: pytorch_lightning/logging/neptune.py (hunk 4 above)
Comment (Borda) [reply pytorch_lightning/logging/neptune.py]: well you can make a simple fork via using single variable:
```
def append_tags(self, tags):
    if not isinstance(tags, (list, set, tuple)):
        tags = [tags]  # make it as an iterable is if it is not yet
    ...
```

**Comment by Borda** [comment]
> 3 travis jobs worked but 2 still [fail](https://travis-ci.org/williamFalcon/pytorch-lightning/jobs/629639792?utm_medium=notification&utm_source=github_status#L698).
> Could you please take a look @Borda

it looks strange, it seems that the checkpoint is not created, can you can the test locally?

**Comment by jakubczakon** [comment]
Locally it passes and in those 3 other travis jobs it passes as well.

File: pytorch_lightning/logging/neptune.py (hunk 5 above)
Comment (Borda) [reply pytorch_lightning/logging/neptune.py]: this is so more parameters in the docstring...

Comment (jakubczakon) [reply pytorch_lightning/logging/neptune.py]: I think added spaces wherever needed.

Do you mean I should refactor passing neptune-specific parameters as one dictionary or something like that?

**Comment by kuynzereb** [comment]
I have one hypothesis. Default checkpoint is initialized with `save_top_k=1`, so it saves only one checkpoint. And this checkpoint may be not from the last epoch, but the test tries to find the checkpoint from the last epoch

**Comment by kuynzereb** [comment]
I have had the similar issue in #653 and setting `save_top_k=-1` really helped.

**Comment by jakubczakon** [comment]
Thank you @kuynzereb and @Borda it worked!

**Comment by Borda** [comment]
> I have one hypothesis. Default checkpoint is initialized with `save_top_k=1`, so it saves only one checkpoint. And this checkpoint may be not from the last epoch, but the test tries to find the checkpoint from the last epoch

true, in theory, it may happen that the next epoch does not lower training/validation loss as the loss is usually oscillating and the test case is to be minimal in terms of time and computational power...

**Comment by jakubczakon** [comment]
Hi there @Borda, happy new years!

I was wondering if there is anything else I could do to help merge this PR?

**Comment by Borda** [comment]
@jakubczakon nop, it is good from my side, Nice work!
@williamFalcon could you have look, it LGTM

**Comment by williamFalcon** [comment]
@jakubczakon great addition!

**Pull request merged**